app/src/main/java/com/rootstrap/android/ui/activity/main/SignInActivityViewModel.kt

```diff
@@ -39,22 +42,38 @@ open class SignInActivityViewModel(
         viewModelScope.launch {
             try {
                 val result = manager.signIn(user = user)
-                if (result.isSuccess) {
-                    result.getOrNull()?.value?.user?.let { user ->
-                        SessionManager.signIn(user)
-                    }
+                signInResult(result)
+            } catch (exception: IOException) {
+                handleError(Throwable())
+            }
+        }
+    }
 
-                    networkState = NetworkState.idle
-                    state = SignInState.signInSuccess
-                } else {
-                    handleError(result.exceptionOrNull())
-                }
+    fun signInWithFacebook(token: String) {
+        networkState = NetworkState.loading
+        SessionManager.addFacebookToken(token)
+        viewModelScope.launch {
+            try {
+                val result = manager.signInWithFacebook(FacebookSignIn(token))
+                signInResult(result)
             } catch (exception: IOException) {
                 handleError(Throwable())
             }
         }
     }
 
+    private fun signInResult(result: Result<Data<UserSerializer>>) {
+        if (result.isSuccess) {
+            result.getOrNull()?.value?.user?.let { user ->
+                SessionManager.signIn(user)
+            }
+            networkState = NetworkState.idle
+            state = SignInState.signInSuccess
```

Comment:
you should add a new sign in state for Facebook(facebookSignInSucess), because the is different from the normal signIn, for now, it's ok but keep in mind for the future.

Reply:
Thank you for the advice